internal/zinc-persist/src/main/scala/sbt/internal/inc/binary/converters/ProtobufWriters.scala

```diff
@@ -323,33 +308,26 @@ final class ProtobufWriters(mapper: WriteMapper) {
         .setValue(annotationArgument.value())
         .build
 
-    val arguments = annotation.arguments().toSchemaList(toAnnotationArgument)
     val base = toType(annotation.base())
-    Schema.Annotation.newBuilder
-      .setBase(base)
-      .addAllArguments(arguments.asJava)
-      .build
+    val builder = Schema.Annotation.newBuilder.setBase(base)
+    annotation.arguments().foreach(a => builder.addArguments(toAnnotationArgument(a)))
+    builder.build
   }
 
   def toStructure(tpe: Structure): Schema.Type.Structure = {
-    val declared = tpe.declared().toSchemaList(toClassDefinition)
-    val inherited = tpe.inherited().toSchemaList(toClassDefinition)
-    val parents = tpe.parents().toSchemaList(toType)
-    Schema.Type.Structure.newBuilder
-      .addAllDeclared(declared.asJava)
-      .addAllInherited(inherited.asJava)
-      .addAllParents(parents.asJava)
-      .build
+    val builder = Schema.Type.Structure.newBuilder
+    tpe.declared().foreach(t => builder.addDeclared(toClassDefinition(t)))
+    tpe.inherited().foreach(i => builder.addInherited(toClassDefinition(i)))
+    tpe.parents().foreach(p => builder.addParents(toType(p)))
+    builder.build
```

Comment:
Thanks for this! In the hindsight 2020 it makes sense that `.asJava` that I added in https://github.com/sbt/zinc/pull/799 was suboptimal.

Reply:
Teamwork!

In the absence of automated performance testing, it seems inevitable that this kind of performance regression might occur. I wouldn't have noticed the regression were it not for https://github.com/eatkins/scala-build-watch-performance. That being said, a useful heuristic that I try to follow is to avoid copying data if at all possible. It's often hard to tell in isolation whether or not copying will impact performance. I also try to practice the boy scout rule of leaving things better than I found them.